sonar-batch/src/main/java/org/sonar/batch/report/TestExecutionAndCoveragePublisher.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * SonarQube, open source software quality management tool.
+ * Copyright (C) 2008-2014 SonarSource
+ * mailto:contact AT sonarsource DOT com
+ *
+ * SonarQube is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * SonarQube is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.batch.report;
+
+import com.google.common.base.Function;
+import com.google.common.collect.Iterables;
+import org.sonar.api.batch.fs.InputFile.Type;
+import org.sonar.api.batch.fs.internal.DefaultInputFile;
+import org.sonar.api.test.CoverageBlock;
+import org.sonar.api.test.MutableTestCase;
+import org.sonar.api.test.MutableTestPlan;
+import org.sonar.api.test.TestCase;
+import org.sonar.batch.index.BatchResource;
+import org.sonar.batch.index.ResourceCache;
+import org.sonar.batch.protocol.Constants.TestStatus;
+import org.sonar.batch.protocol.output.BatchReport;
+import org.sonar.batch.protocol.output.BatchReport.CoverageDetail;
+import org.sonar.batch.protocol.output.BatchReport.Test;
+import org.sonar.batch.protocol.output.BatchReportWriter;
+import org.sonar.core.test.TestPlanBuilder;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class TestExecutionAndCoveragePublisher implements ReportPublisherStep {
+
+  private final class FunctionImplementation implements Function<String, CoverageDetail> {
+    private final MutableTestPlan testPlan;
+    private BatchReport.CoverageDetail.Builder builder = BatchReport.CoverageDetail.newBuilder();
+    private BatchReport.CoverageDetail.CoveredFile.Builder coveredBuilder = BatchReport.CoverageDetail.CoveredFile.newBuilder();
+
+    private FunctionImplementation(MutableTestPlan testPlan) {
+      this.testPlan = testPlan;
+    }
+
+    @Override
+    public CoverageDetail apply(String testName) {
+      // Take first test with provided name
+      MutableTestCase testCase = testPlan.testCasesByName(testName).iterator().next();
```

Comment:
:new: ![CRITICAL](https://raw.githubusercontent.com/henryju/image-hosting/master/critical.png)testName must be non-null but is marked as nullable[![...](https://raw.githubusercontent.com/henryju/image-hosting/master/more.png)](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_PARAMETER_MUST_BE_NONNULL_BUT_MARKED_AS_NULLABLE)